File: src/main/java/com/arcao/geocaching4locus/error/exception/IntendedException.java
package com.arcao.geocaching4locus.error.exception;

import android.content.Intent;
import com.arcao.geocaching4locus.base.util.ParcelableUtil;
import java.io.IOException;

public class IntendedException extends Exception {
  private static final long serialVersionUID = -6278705769679870918L;
  private transient Intent intent;

  public IntendedException(Throwable throwable, Intent intent) {
    super(throwable);
    this.intent = intent;
  }

  public Intent getIntent() {
    return intent;
  }

  private void writeObject(java.io.ObjectOutputStream out) throws IOException {
    out.defaultWriteObject();

    byte[] data = ParcelableUtil.marshall(intent);

    out.writeInt(data.length);
    out.write(data);
  }
  private void readObject(java.io.ObjectInputStream in) throws IOException, ClassNotFoundException {
    in.defaultReadObject();

    int len = in.readInt();
    byte[] data = new byte[len];

    if (in.read(data) != len)
      return;

    intent = ParcelableUtil.unmarshall(data, Intent.CREATOR);
  }
}

File: app/src/main/java/com/arcao/geocaching4locus/error/exception/IntendedException.kt
package com.arcao.geocaching4locus.error.exception

import android.content.Intent
import com.arcao.geocaching4locus.base.util.marshall
import com.arcao.geocaching4locus.base.util.unmarshall
import java.io.IOException
import java.io.ObjectInputStream
import java.io.ObjectOutputStream

class IntendedException(throwable: Throwable, intent: Intent) : Exception(throwable) {
    @Transient
    var intent: Intent = intent
        private set

    @Throws(IOException::class)
    private fun writeObject(out: ObjectOutputStream) {
        out.defaultWriteObject()

        val data = intent.marshall()
        out.writeInt(data.size)
        out.write(data)
    }

    @Throws(IOException::class, ClassNotFoundException::class)
    private fun readObject(inputStream: ObjectInputStream) {
        inputStream.defaultReadObject()

        val len = inputStream.readInt()
        val data = ByteArray(len)

        if (inputStream.read(data) != len) throw IOException("Corrupted Parcelable data")
        intent = data.unmarshall(Intent.CREATOR)
    }

    companion object {
        private const val serialVersionUID = -6278705769679870918L
    }
}
